evm/vm_state.py

```diff
@@ -0,0 +1,299 @@
+        # further modifications can occur using the `State` object after
+        # leaving the context.
+        state.db = None
+        state._trie = None
+
+    #
+    # Snapshot and Revert
+    #
+    def snapshot(self):
+        """
+        Perform a full snapshot of the current state.
+
+        Snapshots are a combination of the state_root at the time of the
+        snapshot and the checkpoint_id returned from the journaled DB.
+        """
+        return (self.block_header.state_root, self._chaindb.snapshot())
+
+    def revert(self, snapshot):
+        """
+        Revert the VM to the state at the snapshot
+        """
+        state_root, checkpoint_id = snapshot
+
+        with self.state_db() as state_db:
+            # first revert the database state root.
+            state_db.root_hash = state_root
+            # now roll the underlying database back
+
+        self._chaindb.revert(checkpoint_id)
+
+    def commit(self, snapshot):
+        """
+        Commits the journal to the point where the snapshot was taken.  This
+        will destroy any journal checkpoints *after* the snapshot checkpoint.
+        """
+        _, checkpoint_id = snapshot
+        self._chaindb.commit(checkpoint_id)
+
+    #
+    # Access ChainDB (Read-only)
+    #
+    def get_ancestor_hash(self, block_number):
+        """
+        Return the hash for the ancestor with the given block number.
+        """
+        ancestor_depth = self.block_header.block_number - block_number
+        if ancestor_depth > 256 or ancestor_depth < 1:
+            return b''
+        header = self._chaindb.get_block_header_by_hash(self.block_header.parent_hash)
+        while header.block_number != block_number:
+            header = self._chaindb.get_block_header_by_hash(header.parent_hash)
+        return header.hash
+
+    def get_block_header_by_hash(self, block_hash):
+        """
+        Returns the block header by hash.
+        """
+        return self._chaindb.get_block_header_by_hash(block_hash)
+
+    def get_parent_header(self, block_header):
+        """
+        Returns the header for the parent block.
+        """
+        return self.get_block_header_by_hash(block_header.parent_hash)
+
+    def is_key_exsits(self, key):
+        """
+        Check if the given key exsits in chaindb
+        """
+        return self._chaindb.exists(key)
+
+    #
+    # Computation
+    #
+    def get_computation(self, message):
+        """Return state object
+        """
+        if self.computation_class is not None:
+            computation = self.computation_class(self, message)
+        else:
+            raise AttributeError("No `computation_class` has been set for this VMState")
+        return computation
+
+    #
+    # Execution
+    #
+    @classmethod
+    def apply_transaction(
+            cls,
+            vm_state,
+            transaction,
+            block,
+            is_stateless=True,
```

Comment:
I think there is a route for us to remove the need for this flag.  We can have `VM.apply_transaction` **always** be pure/stateless, and then modify `Chain.apply_transaction` to update the local vm state before returning the `computation` object.

Assuming there is nothing wrong with that approach I like it a lot more than using a flag.

Reply:
Agree! Now the non-stateless mode is only for verifying the stateless mode result in case I screw up. Could we remove the non-stateless mode later?

Reply:
Yes, doing it separately is fine.  Mind opening an issue to track it and linking it here.